netconf/man/netconf-subsystem.1
.\" Process this file with
.\" nroff -e -mandoc foo.1
.\"
.TH netconf-subsystem 1 "March 23, 2025" Linux "netconf-subsystem 2.15"
.SH NAME
netconf-subsystem \- thin client to access netconfd server

.SH SYNOPSIS
.nf

   netconf-subsystem

.fi
.SH DESCRIPTION
.B netconf-subsystem
is a thin-client application that is called by the
OpenSSH server to access the netconfd server, when 
the 'netconf' subsystem is requested.
.SH USAGE
The location of this program needs to be configured
in the /etc/ssh/sshd_config file.  The following
configuration needs to be present:
.nf

  Port 22
  Port 830
  Subsystem netconf /usr/sbin/netconf-subsystem

.fi
The actual filespec in the last line will depend
on the location that this program is installed.
The default value is shown in the example above.

.SH OPTIONS
.IP --\fBncxserver-sockname\fP=port@filespec
Specifies the corresponding pairs of SSH ports
and UNIX socket names. This option allows multiple
netconfd instances to run on the same host.

/etc/ssh/sshd_config:
.nf
\&...
Port 830
Port 1830
Port 2830
Subsystem netconf "/usr/sbin/netconf-subsystem \
--ncxserver-sockname=830@/tmp/ncxserver.sock \
--ncxserver-sockname=1830@/tmp/ncxserver-right.sock \
--ncxserver-sockname=2830@/tmp/ncxserver-middle.sock"
\&...
.fi

.SH AUTHORS
Andy Bierman, <andy at netconfcentral dot org>

Vladimir Vassilev, <vladimir at lightside-instruments dot com>
.SH SEE ALSO
.BR netconfd (1)


CHANGELOG
yuma123 (2.15) UNRELEASED; urgency=medium

  * 

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Tue, 20 May 2025 14:28:55 +0200

yuma123 (2.14) stable; urgency=medium

  * Added support for notifications defined inside list
  * Bug in cli parsing list keys containing spaces fixed
  * Added agt_fd_event_cb_register(fd, callback) function for file descriptor event interrupts
  * Fixed bug in --dump-session always reporting of received chunk size 0 in *-in.ts
  * Fixed bug in val_set_cplxval_obj function always returning NO_ERR even if validation fails

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Sun, 23 Mar 2025 15:25:36 +0100

yuma123 (2.13) stable; urgency=medium

  * ncx
    - Added if-feature parsing support for YANG 1.1 logical expressions
    - Fixed bug in obj_next_child_deep processing of models with multiple
      levels of choice/case statements
    - Fixed bug in ncx_sprintf_num for decimal64 e.g. -0.000068623 was
      represented as 0.0000-68623
  * netconfd
    - Fixed bug causing netconfd to not handle --running-error=continue
  * yangcli
    - Fill for rpc issue solved by not serializing the virtual <input>
    - Fixed socket connection timeout and signal handlng e.g. ctrl-c now
      terminates stall connection attempt
  * autoconf
    - pkg-config is now used to find libxml2
  * python/test
    - transition to python3
  * examples
    - Added initial ietf-ip support as part of the ietf-interfaces example

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Wed, 30 Nov 2022 17:20:00 +0100

yuma123 (2.12) stable; urgency=medium

  * Added val123_add_virtual_cb to support multiple calbacks "filling" common
    virtual container.
  * Fixed anyxml support for config=true data
  * Improved /system-state/clock/current-datetime resolution to nanoseconds and
    disabled value caching
  * Augmented set-my-session and get-my-session RPCs with cache-timeout param
    allowing cache to be disabled when cache-timeout=0
  * Fixed Xpath compare of nodeset to nodeset implementation for numeric nodesets
  * Updated IETF draft YANG modules

 -- Vladimir Vassilev <vladimir@lightside-instruments.com>  Wed, 18 Aug 2021 00:08:23 +0200

yuma123 (2.11) stable; urgency=medium

  * Improved stability and interoperability
  * Added more validation testcases
  * Preliminary NMDA support rfc8342
  * RPM build scripts (from ekinzie)
  * yangcli
    - Added --use-agent parm for use of ssh-agent authentication (from ekinzie)
  * yangdump
    - YANG tree format support rfc8340

 -- Vladimir Vassilev <vladimir@transpacket.com>  Fri, 17 Aug 2018 03:02:54 +0200

yuma123 (2.10) stable; urgency=medium

  * ncx
    - YANG 1.1 Xpath functions implementation.
    - Fixed bug in handling of deviations.
    - Fixed bug in apply_commit_deletes.
    - Fixed bug in val_make_serialized_string.
    - Refactored the base64 implementation. Fixed bugs and inconsistencies.
    - Added support for multiple obj callback registrations.
    - Fixed CDATA handling.
    - Splitted yuma-netconf.yang to yuma123-netconf.yang and yuma123-netconf-types.yang
    - Various ietf draft models updated. Erratas applied.
    - Removed circular libyumancx dependencies on libyumamgr
  * agt
    - Implementation of ietf-yang-library
    - Replaced proprietary notifications with the ietf-netconf-notifications.yang defined.
    - Moved agt_arp.c integrated module as standalone example-modules/yuma-arp
  * test
    - Added netconfd testsuite netconf/test/netconfd.
    - Added litenc client side session scripting python module.
  * yangcli
    - Implemented ietf-yang-library client side support.
    - Fixed bug caused by attempted reload of already loaded module.
    - Added yes/no container fill check promp.

 -- Vladimir Vassilev <vladimir@transpacket.com>  Fri, 29 Sep 2017 13:53:28 +0200

yuma123 (2.9) stable; urgency=medium

  * Removed IETF copyrighted MIBs to avoid the complications related to +dfsg releases
  * yangrpc API refactored for stability allowing new options passed as yangcli  parameter string
  * Added dump-session yangcli parameter

 -- Vladimir Vassilev <vladimir@transpacket.com>  Sat, 20 Aug 2016 11:46:34 +0200

yuma123 (2.8+dfsg) stable; urgency=medium

  * Removed IETF copyrighted MIBs and YANG models derived from them

 -- Vladimir Vassilev <vladimir@transpacket.com>  Sun, 07 Aug 2016 15:17:40 +0200

yuma123 (2.8) stable; urgency=medium

  * Resolved bobhash license-problem-non-free-RFC-BCP78
  * Fixed unescaped dots in man pages
  * Fixed debian lintian warning out-of-date-standards-version
  * Separated libyangrpc from the libyuma package

 -- Vladimir Vassilev <vladimir@transpacket.com>  Thu, 28 Jul 2016 23:19:24 +0200

yuma123 (2.7-1) stable; urgency=medium

  * Cleaned up debian
  * Renamed yuma libraries libncx->libyumancx, libagt->libyumaagt, libmgr->libyumamgr

 -- Vladimir Vassilev <vladimir@transpacket.com>  Thu, 21 Jul 2016 18:16:16 +0200

yuma123 (2.6-1) stable; urgency=medium

  * ncx
    - Added val_make_serialized_string to the API
  * netconfd
    - Fixed a bug caused by system time adjustments and the use of nonmonotonic time function for timer callback implementation
    - Fixed bugs both in netconf 1.0 and netconf 1.1 message parsers
    - Added readline as default cli interface library with option to build with tecla instead
    - Fixed bugs in must statement validation
  * yangcli
    - Added readline as default cli interface library with option to build with tecla instead
  * examples
    - Added standalone automake based example projects building ietf-interfaces.yang and ietf-system.yang implementations for linux
  * yangrpc
    - Added yangrpc_parse_cli API parsing yangcli style command lines based on the schema context of the connection
  * python/perl
    - Initial perl and python bindings

 --  <vladimir@transpacket.com>  Sun, 17 Jul 2016 10:10:00 +0200

yuma123 (2.5-1) stable; urgency=low

  * Implemented direct-must-augment-ex:augment extension


netconf/src/subsys/netconf-subsystem.c
/*
 * Copyright (c) 2008 - 2012, Andy Bierman, All Rights Reserved.
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 */
/*  FILE: netconf-subsystem.c

                
*********************************************************************
*                                                                   *
*                  C H A N G E   H I S T O R Y                      *
*                                                                   *
*********************************************************************

date         init     comment
----------------------------------------------------------------------
14-jan-07    abb      begun;
03-mar-11    abb      get rid of usleeps and replace with
                      design that checks for EAGAIN

*********************************************************************
*                                                                   *
*                     I N C L U D E    F I L E S                    *
*                                                                   *
*********************************************************************/
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <netdb.h>
#include <unistd.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <pwd.h>
#include <stdarg.h>

#define _C_main 1

#include "procdefs.h"

#ifndef _H_ncxconst
#include "ncxconst.h"
#endif

#ifndef _H_agt
#include "agt.h"
#endif

#ifndef _H_agt_ncxserver
#include "agt_ncxserver.h"
#endif

#ifndef _H_status
#include "status.h"
#endif

#ifndef _H_send_buff
#include "send_buff.h"
#endif

#ifndef _H_xml_util
#include "xml_util.h"
#endif


/********************************************************************
*                                                                   *
*                       C O N S T A N T S                           *
*                                                                   *
*********************************************************************/
#define BUFFLEN  2000

#define MAX_READ_TRIES 1000

/********************************************************************
*                                                                   *
*                       V A R I A B L E S                           *
*                                                                   *
*********************************************************************/

static char *client_addr;
static struct sockaddr *ncxname;
static struct sockaddr_un ncxname_unix;
static struct sockaddr_in ncxname_inet;
static char*    ncxserver_sockname;
static int    ncxport_inet;
static int    ncxsock;
static char *user;
static char *port;

static int traceLevel = 0;
static FILE *errfile;

#define SUBSYS_TRACE1(fmt, ...) if (traceLevel && errfile) \
                                { \
                                    fprintf( errfile, fmt, ##__VA_ARGS__); \
                                    fflush( errfile ); \
                                }

#define SUBSYS_TRACE2(fmt, ...) if (traceLevel > 1 && errfile) \
                                { \
                                    fprintf( errfile, fmt, ##__VA_ARGS__); \
                                    fflush( errfile ); \
                                }

#define SUBSYS_TRACE3(fmt, ...) if (traceLevel > 2 && errfile) \
                                { \
                                    fprintf( errfile, fmt, ##__VA_ARGS__); \
                                    fflush( errfile ); \
                                }

static boolean ncxconnect;
static char msgbuff[BUFFLEN];

/******************************************************************
 * FUNCTION configure_logging
 *
 * Configure debug logging. This function evaluates command line
 * arguments to configure debug logging.
 *
 ******************************************************************/
static void configure_logging( int argc, char** argv )
{
    char* err_filename;
    int arg_idx = 1;
    char  defname[21];

    strncpy(defname, "/tmp/subsys-err.log", 20);
    err_filename = defname;
    while ( arg_idx < argc-1 ) {
        if ( !strcmp( argv[arg_idx], "-filename" ) ||
             !strcmp( argv[arg_idx], "-f" ) ) {
            err_filename = argv[++arg_idx];
            if ( !traceLevel )
            {
                traceLevel = 1;
            }
        }
        else if ( !strcmp( argv[arg_idx], "-trace" ) ||
                  !strcmp( argv[arg_idx], "-t" ) ) {
            traceLevel = atoi( argv[++arg_idx] );
        }
        ++arg_idx;
    }

    if ( traceLevel ) {
        errfile = fopen( err_filename, "a" );
        SUBSYS_TRACE1( "\n*** New netconf Session Started ***\n" );
    }

}

/********************************************************************
* FUNCTION init_subsys
*
* Initialize the subsystem, and get it ready to send and receive
* the first message of any kind
* 
* RETURNS:
*   status
*********************************************************************/
static status_t
    init_subsys (int argc, char** argv)
{
    char *cp, *con;
    int   ret;
    int name_size;
    int i;

    client_addr = NULL;
    port = NULL;
    user = NULL;
    ncxsock = -1;
    ncxconnect = FALSE;
    ncxport_inet = -1;
    ncxserver_sockname = NULL;

    for(i=1;i<argc;i++) {
    	if(strlen(argv[i])>strlen("--tcp-direct-port=") && 0==memcmp(argv[i],"--tcp-direct-port=",strlen("--tcp-direct-port="))) {
            ncxport_inet = atoi(argv[i]+strlen("--tcp-direct-port=")); 
        } else if(strlen(argv[i])>strlen("--ncxserver-sockname=") && 0==memcmp(argv[i],"--ncxserver-sockname=",strlen("--ncxserver-sockname="))) {
            ncxserver_sockname = argv[i]+strlen("--ncxserver-sockname=");
    	}
    }    

    /* get the client address */
    con = getenv("SSH_CONNECTION");
    if (!con) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): "
                       "Get SSH_CONNECTION variable failed\n" );
        return ERR_INTERNAL_VAL;
    }

    /* get the client addr */
    client_addr = strdup(con);
    if (!client_addr) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): strdup(client_addr) failed\n" );
        return ERR_INTERNAL_MEM;
    }
    cp = strchr(client_addr, ' ');
    if (!cp) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): "
                       "Malformed SSH_CONNECTION variable\n" );
        return ERR_INTERNAL_VAL;
    } else {
        *cp = 0;
    }

    /* get the server connect port */
    cp = strrchr(con, ' ');
    if (cp && cp[1]) {
        port = strdup(++cp);
    }
    if (!port) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): "
                       "Malformed SSH_CONNECTION variable\n" );
        return ERR_INTERNAL_VAL;
    }
        
    /* get the username */
    cp = getenv("USER");
    if (!cp) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): Get USER variable failed\n");
        return ERR_INTERNAL_VAL;
    }
    user = strdup(cp);
    if (!user) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): strdup(user) failed\n" );
        return ERR_INTERNAL_MEM;
    }
    if(ncxport_inet != -1)
    {
        struct hostent* hp;
        int tcp_nodelay_option=1;

        /* make a socket to connect to the NCX server */
        ncxsock = socket(AF_INET, SOCK_STREAM, 0);
        if (ncxsock < 0) {
            SUBSYS_TRACE1( "ERROR: init_subsys(): NCX Socket Creation failed\n" );
            return ERR_NCX_CONNECT_FAILED;
        }
	if (setsockopt(ncxsock, IPPROTO_TCP, TCP_NODELAY, (char*) &tcp_nodelay_option, sizeof(tcp_nodelay_option)) < 0) {
            SUBSYS_TRACE1( "ERROR: init_subsys(): NCX Socket Creation failed\n" );
            return ERR_NCX_CONNECT_FAILED;
        }
        hp = gethostbyname("localhost");
        if (hp == NULL) {
            SUBSYS_TRACE1( "ERROR: init_subsys(): NCX Socket Creation failed\n" );
            return ERR_NCX_CONNECT_FAILED;
        }
        memset((char *) &ncxname_inet, 0, sizeof(ncxname_inet));
        ncxname_inet.sin_family = AF_INET;
        ncxname_inet.sin_port = htons((unsigned short)ncxport_inet);
        memcpy((char *) &ncxname_inet.sin_addr, hp->h_addr, hp->h_length);
        
        name_size = sizeof(ncxname_inet);
        ncxname = (struct sockaddr *) &ncxname_inet;
    } else {
        /* make a socket to connect to the NCX server */
        ncxsock = socket(PF_LOCAL, SOCK_STREAM, 0);
        if (ncxsock < 0) {
            SUBSYS_TRACE1( "ERROR: init_subsys(): NCX Socket Creation failed\n" );
            return ERR_NCX_CONNECT_FAILED;
        } 
        ncxname_unix.sun_family = AF_LOCAL;
        strncpy(ncxname_unix.sun_path, 
                (ncxserver_sockname==NULL)?NCXSERVER_SOCKNAME:ncxserver_sockname, 
                sizeof(ncxname_unix.sun_path));
        name_size = SUN_LEN(&ncxname_unix);
        ncxname = (struct sockaddr *)&ncxname_unix;
    }

    /* try to connect to the NCX server */
    ret = connect(ncxsock,
                  ncxname,
                  name_size);
    if (ret != 0) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): NCX Socket Connect failed\n" );
        return ERR_NCX_OPERATION_FAILED;
    } else {
        SUBSYS_TRACE2( "INFO:  init_subsys(): "
                       "NCX Socket Connected on FD: %d \n", ncxsock );
        ncxconnect = TRUE;
    }

#ifdef USE_NONBLOCKING_IO
    /* set non-blocking IO */
    if (fcntl(ncxsock, F_SETFD, O_NONBLOCK)) {
        SUBSYS_TRACE1( "ERROR: init_subsys(): fnctl() failed\n" );
    }
#endif

    /* connected to the ncxserver and setup the ENV vars ok */
    return NO_ERR;

} /* init_subsys */


/********************************************************************
* FUNCTION cleanup_subsys
*
* Cleanup the subsystem
* 
*********************************************************************/
static void
    cleanup_subsys (void)
{
    if (client_addr) {
        m__free(client_addr);
    }
    if (user) {
        m__free(user);
    }
    if (ncxconnect) {
        close(ncxsock);
    }

    if (errfile) {
        fclose(errfile);
    }
} /* cleanup_subsys */


/********************************************************************
* FUNCTION send_ncxconnect
*
* Send the <ncx-connect> message to the ncxserver
* 
* RETURNS:
*   status
*********************************************************************/
static status_t
    send_ncxconnect (void)
{
    status_t  res;
    char connectmsg[] = 
    "%s\n<ncx-connect xmlns=\"%s\" version=\"1\" user=\"%s\" "
    "address=\"%s\" magic=\"%s\" transport=\"ssh\" port=\"%s\" />\n%s";

    memset(msgbuff, 0x0, BUFFLEN);
    snprintf(msgbuff, BUFFLEN, connectmsg, (const char *)XML_START_MSG, 
             NCX_URN, user, client_addr, NCX_SERVER_MAGIC, port, NC_SSH_END);

    res = send_buff(ncxsock, msgbuff, strlen(msgbuff));
    return res;

} /* send_ncxconnect */


/********************************************************************
* FUNCTION do_read
*
* Read from a FD
* 
* INPUTS:
*              
* RETURNS:
*   return byte count
*********************************************************************/
static ssize_t
    do_read (int readfd, 
             char *readbuff, 
             size_t readcnt, 
             status_t *retres)
{
    boolean   readdone;
    ssize_t   retcnt;

    readdone = FALSE;
    retcnt = 0;
    *retres = NO_ERR;

    while (!readdone && *retres == NO_ERR) {
        retcnt = read(readfd, readbuff, readcnt);
        if (retcnt < 0) {
            if (errno != EAGAIN) {
                SUBSYS_TRACE1( "ERROR: do_read(): read of FD(%d): "
                               "failed with error: %s\n", 
                               readfd, strerror( errno  ) );
                *retres = ERR_NCX_READ_FAILED;
                continue;
            }
        } else if (retcnt == 0) {
            SUBSYS_TRACE1( "INFO: do_read(): closed connection\n");
            *retres = ERR_NCX_EOF;
            readdone = TRUE;
            continue;
        } else {
            /* retcnt is the number of bytes read */
            readdone = TRUE;
        }
    }  /*end readdone loop */

    return retcnt;

}  /* do_read */


/********************************************************************
* FUNCTION io_loop
*
* Handle the IO for the program
* 
* INPUTS:
*              
* RETURNS:
*   status
*********************************************************************/
static status_t
    io_loop (void)
{
    status_t  res;
    boolean   done;
    fd_set    fds;
    int       ret;
    ssize_t   retcnt;

    res = NO_ERR;
    done = FALSE;
    FD_ZERO(&fds);

    while (!done) {
        FD_SET(STDIN_FILENO, &fds);
        FD_SET(ncxsock, &fds);

        ret = select(FD_SETSIZE, &fds, NULL, NULL, NULL);
        if (ret < 0) {
            if ( errno != EINTR ) {
                SUBSYS_TRACE1( "ERROR: io_loop(): select() "
                               "failed with error: %s\n", strerror( errno ) );
                res = ERR_NCX_OPERATION_FAILED;
                done = TRUE;
            }
            else
            {
                SUBSYS_TRACE2( "INFO: io_loop(): select() "
                               "failed with error: %s\n", strerror( errno  ) );
            }
            continue;
        } else if (ret == 0) {
            SUBSYS_TRACE1( "ERROR: io_loop(): select() "
                           "returned 0, exiting...\n" );
            res = NO_ERR;
            done = TRUE;
            continue;
        } /* else some IO to process */

        /* check any input from client */
        if (FD_ISSET(STDIN_FILENO, &fds)) {
            /* get buff from openssh */
            retcnt = do_read(STDIN_FILENO, 
                             msgbuff, 
                             (size_t)BUFFLEN,
                             &res);

            if (res == ERR_NCX_EOF) {
                res = NO_ERR;
                done = TRUE;
                continue;
            } else if (res == ERR_NCX_SKIPPED) {
                res = NO_ERR;
            } else if (res == NO_ERR && retcnt > 0) {
                /* send this buffer to the ncxserver */
                res = send_buff(ncxsock, msgbuff, (size_t)retcnt);
                if (res != NO_ERR) {
                    SUBSYS_TRACE1( "ERROR: io_loop(): send_buff() to ncxserver "
                                   "failed with %s\n", strerror( errno ) );
                    done = TRUE;
                    continue;
                }
            }
        }  /* if STDIN set */

        /* check any input from the ncxserver */
        if (FD_ISSET(ncxsock, &fds)) {
            res = NO_ERR;
            retcnt = do_read(ncxsock, 
                             msgbuff, 
                             (size_t)BUFFLEN,
                             &res);

            if (res == ERR_NCX_EOF) {
                res = NO_ERR;
                done = TRUE;
                continue;
            } else if (res == ERR_NCX_SKIPPED) {
                res = NO_ERR;
            } else if (res == NO_ERR && retcnt > 0) {
                /* send this buffer to STDOUT */
                res = send_buff(STDOUT_FILENO, msgbuff, (size_t)retcnt);
                if (res != NO_ERR) {
                    SUBSYS_TRACE1( "ERROR: io_loop(): send_buff() to client "
                                   "failed with %s\n", strerror( errno ) );
                    done = TRUE;
                    continue;
                }
            }
        }
    }

    return res;

} /* io_loop */


/********************************************************************
* FUNCTION main
*
* STDIN is input from the SSH client (sent to ncxserver)
* STDOUT is output to the SSH client (rcvd from ncxserver)
* 
* RETURNS:
*   0 if NO_ERR
*   1 if error connecting or logging into ncxserver
*********************************************************************/
int main (int argc, char **argv) 
{
    status_t  res;
    const char *msg;

    configure_logging( argc, argv );

    res = init_subsys(argc, argv);
    if (res != NO_ERR) {
        msg = "init failed";
    }

    if (res == NO_ERR) {
        res = send_ncxconnect();
        if (res != NO_ERR) {
            msg = "connect failed";
        }
    }

    if (res == NO_ERR) {
        res = io_loop();
        if (res != NO_ERR) {
            msg = "IO error";
        }
    }

    if (res != NO_ERR) {
        SUBSYS_TRACE1( "ERROR: io_loop(): exited with error %s \n", msg );
    }

    cleanup_subsys();

    if (res != NO_ERR) {
        return 1;
    } else {
        return 0;
    }

} /* main */

